### PR TITLE
[debugger] Assert when async debug a generic method

### LINKED ISSUE #17549: Debugging 'async Task<T> Name<T>()' method makes app crash
_From @mykyta-bondarenko on Thu, 24 Oct 2019 13:42:52 GMT_

### Steps to Reproduce

1. Download a [sample](https://github.com/xamarin/xamarin-macios/files/3767795/async-debugging-bug.zip) 
2. Open the project and `ViewController.cs` file 
3. Add a breakpoint to 20th line (`await Task.Delay(2);`)
4. Press `F5`
5. Press `F10` when the debugger will be on that line 

### Expected Behavior
The debugger works as expected 

### Actual Behavior
The debugger and app crash

### Environment

1. VSfM
https://gist.github.com/mykyta-bondarenko/48b11cb05c6f4c1e80a7f8827c1d174f
2. Code

```
        public override async void ViewDidLoad()
        {
            base.ViewDidLoad();
            await ExecuteAsync_Broken<object>();
        }

        async Task<T> ExecuteAsync_Broken<T>()
        {
            await Task.Delay(2);
            return default;
        }
```

### Build Logs

1. Application Output
https://gist.github.com/mykyta-bondarenko/6b4ba6f441c43d6d28000559a8c85f48
2. Build Output
https://gist.github.com/mykyta-bondarenko/239e0e45616e81a3c27bc12f3f94a640

### Example Project 
[Sample](https://github.com/xamarin/xamarin-macios/files/3767795/async-debugging-bug.zip)

_Copied from original issue xamarin/xamarin-macios#7296_

### LINKED ISSUE #17569: Await within method with return type Task<T> makes debugging fail
## Steps to Reproduce

1. Download and open [AsyncBug-v3.zip](https://github.com/xamarin/xamarin-macios/files/3765057/AsyncBug-v3.zip).
2. Put a breakpoint on App.xaml.cs:46 (the `string url = "https://jsonplaceholder.typicode.com/todos/1";` line).
3. Run on an iOS device.
4. When the breakpoint is hit (there might already be a breakpoint on a similar line further up in the line: make sure you're testing the right breakpoint), hit step-over.

### Current Behavior

After the breakpoint is hit (there may already be a breakpoint on a similar line further up the page), and after doing a 'step-into' or 'step-over', nothing else happens.

### Expected Behavior

Being able to step over the line of code where the breakpoint is; the process should stop on the next line.

## On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:

https://gist.github.com/rolfbjarne/fdb3f0da83e59be22eed42cf25b9f687

## Stacktrace

N/A

### PR DESCRIPTION
**Pull request by thaystg**
When we try to call a method to get the async_id to do an async debug and we are trying to do this in a generic method like this:

```
async Task<T> ExecuteAsync_Broken<T>()
 {
            await Task.Delay(2);
            return default;
  }
```

We need to inflate the generic type before call the method or we will get the error:  Could not execute the method because the containing type 'System.Runtime.CompilerServices.AsyncTaskMethodBuilder1[T_REF]’, is not fully instantiated.

Fixes #17549
Fixes #17569

### CODE REVIEW

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by marek-safar** [comment]
Please add test(s)

File: mono/mini/debugger-agent.h [hunk 1]
```diff
@@ -45,4 +45,7 @@ mono_debugger_agent_stub_init (void);
 MONO_API gboolean
 mono_debugger_agent_transport_handshake (void);
 
+MonoGenericContext
+get_generic_context_from_stack_frame (MonoJitInfo *ji, gpointer generic_info);
```
[the remark this answers is not in the document]
Comment (thaystg) [reply mono/mini/debugger-agent.h]: Done.

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins build Linux WebAssembly

**Comment by thaystg** [comment]
@monojenkins build failed

**Comment by thaystg** [comment]
@monojenkins backport 2019-10

**Pull request merged**